lib/vector-core/buffers/src/disk_v2/ledger.rs

```diff
@@ -0,0 +1,334 @@
+use std::{
+    fmt, io,
+    path::{Path, PathBuf},
+    sync::atomic::{AtomicU16, AtomicU64, Ordering},
```

Comment:
We're doing enough with atomics in this code that'd I'd be real, real interested to get loom involved.

Reply:
I'm not 100% against Loom, although the one thing that might end up being tricky, and I'm not sure how to totally approach it, is that we really don't depend on any of the atomics except for reader/writer file ID, since we primarily use `tokio::sync::Notify` to signal progress between the reader/writer.

Talking through it out loud, though: Tokio's `Notify` doesn't represent any sort of acquire/release synchronization point, so we depend on the atomic ordering when incrementing the reader/writer file IDs, but Loom's `Notify` _does_ simulate an acquire/release synchronization point.

My thought process there is that Loom's `Notify` might give us a stronger guarantee than we would have with the async notifier, and could maybe hide issues... but I'm not sure if that understanding is right _or_ how to quantify that under test with the async behavior.

Reply:
Ah, hmm, that's a fair point. In my imagination for how we'll validate this buffer code we'd need some sort of simulation step -- loom -- to explore any of the re-ordering possibilities and some artificial, hostile filesystem to run a burn-in test on. So, some sort of FUSE thing that makes filesystem operations slow, corrupts data etc. Problems with ordering would turn up in the simulation, overly strong guarantees provided during simulation would show up during burn-in. 

Of course, we'll need to _find_ or build a hostile filesystem.

Reply:
Looks like ScyllaDB came up with a FUSE-based filesystem to do some of this: https://www.scylladb.com/2016/02/16/fault-injection-filesystem-software-testing/

It doesn't address the actual ordering bits, though. 🤔 

I guess a question would be: what's your appetite for long-running burn-in tests replacing a fully-simulated run-through of some of the behavior we use atomics for?  I'm planning on augmenting the builder methods to allow limiting buffer sizes, maximum file counts, etc, which should give us knobs to exercise these cases more often: many more chances to exercise the "has the file rolled over?" logic if files can be filled in 10s of milliseconds instead of seconds, etc.

Reply:
I'd say my appetite is good. My _assumption_ is that if we use simulation methods we will have an easier time debugging and fixing failures versus just burn-in, but there's no reason we can't start with burn-in and go from there at all. So yeah, go as seems good to you and I'll back it.